packages/askar/src/wallet/AskarWallet.ts

```diff
@@ -0,0 +1,745 @@
+          const signed = await signingKeyProvider.sign({
+            data,
+            privateKeyBase58: keyPair.privateKeyBase58,
+            publicKeyBase58: key.publicKeyBase58,
+          })
+
+          return signed
+        }
+        throw new WalletError(`Unsupported keyType: ${key.keyType}`)
+      }
+    } catch (error) {
+      if (!isError(error)) {
+        throw new AriesFrameworkError('Attempted to throw error, but it was not of type Error', { cause: error })
+      }
+      throw new WalletError(`Error signing data with verkey ${key.publicKeyBase58}`, { cause: error })
+    }
+  }
+
+  /**
+   * Verify the signature with the data and the used key
+   *
+   * @param data Buffer The data that has to be confirmed to be signed
+   * @param key Key The key that was used in the signing process
+   * @param signature Buffer The signature that was created by the signing process
+   *
+   * @returns A boolean whether the signature was created with the supplied data and key
+   *
+   * @throws {WalletError} When it could not do the verification
+   * @throws {WalletError} When an unsupported keytype is used
+   */
+  public async verify({ data, key, signature }: WalletVerifyOptions): Promise<boolean> {
+    try {
+      if (!TypedArrayEncoder.isTypedArray(data)) {
+        throw new WalletError(`Currently not supporting signature of multiple messages`)
+      }
+
+      if (keyTypeSupportedByAskar(key.keyType)) {
+        const askarKey = AskarKey.fromPublicBytes({
+          algorithm: keyAlgFromString(key.keyType),
+          publicKey: key.publicKey,
+        })
+        return askarKey.verifySignature({ message: data as Buffer, signature })
+      } else {
+        // Check if there is a signing key provider for the specified key type.
+        if (this.signingKeyProviderRegistry.hasProviderForKeyType(key.keyType)) {
+          const signingKeyProvider = this.signingKeyProviderRegistry.getProviderForKeyType(key.keyType)
+
+          const signed = await signingKeyProvider.verify({
+            data,
+            signature,
+            publicKeyBase58: key.publicKeyBase58,
+          })
+
+          return signed
+        }
+        throw new WalletError(`Unsupported keyType: ${key.keyType}`)
+      }
+    } catch (error) {
+      if (!isError(error)) {
+        throw new AriesFrameworkError('Attempted to throw error, but it was not of type Error', { cause: error })
+      }
+      throw new WalletError(`Error verifying signature of data signed with verkey ${key.publicKeyBase58}`, {
+        cause: error,
+      })
+    }
+  }
+
+  /**
+   * Pack a message using DIDComm V1 algorithm
+   *
+   * @param payload message to send
+   * @param recipientKeys array containing recipient keys in base58
+   * @param senderVerkey sender key in base58
+   * @returns JWE Envelope to send
+   */
+  public async pack(
+    payload: Record<string, unknown>,
+    recipientKeys: string[],
+    senderVerkey?: string // in base58
```

Comment:
Not for now, we should update this to be key instances. The sign/verify already uses keys, and this is just leftover...

Reply:
Also (not for now), but we should update the interface to support DIDComm v2 as well.

Ideally we make the interface more lower level and not `pack`. But rather to encrypt something, and we create a DIDComm packing service that handles the other parts. 

Or we keep it pack but we make the interface more generic and for indy-sdk we only support `xchacha20poly1305_ietf`.

Reply:
I kinda liked how ACA-Py has extracted this into a separate didcomm/v1.py and didcomm/v2.py files to separet the didcomm implementation. Do you think we could do the same?

Reply:
> Ideally we make the interface more lower level and not `pack`. But rather to encrypt something, and we create a DIDComm packing service that handles the other parts.
> 

Yeah, in my opinion this is the way to go: the Wallet provides means to encrypt and sign/verify data while the DIDCommPackingService just uses it (and the DID Resolver) to pack and unpack the message. We'll need of course to define a good interface allowing all the ciphering we are doing for packing.

We can provide in AFJ Core "DIDCommPackingService" for all DIDComm versions and, in the end, supported versions for an Agent will be given by the Wallet it uses.